lib/wallet/descriptor.js

```diff
@@ -0,0 +1,199 @@
+/*!
+ * descriptor.js - descriptor object for bcoin
+ * Copyright (c) 2015-2016, Christopher Jeffrey (MIT License).
+ * https://github.com/bcoin-org/bcoin
+ */
+/* eslint-disable max-len */
+'use strict';
+const assert = require('bsert');
+// const Network = require('../protocol/network');
+const BN = require('bcrypto/lib/bn.js');
+const { checkScriptType, stringAfterScriptType, giveExpr, isHex, isCompressed } = require('../utils/stringparsing');
+const KeyRing = require('../primitives/keyring');
+const { ConstPubkeyProvider } = require('./pubkeyprovider');
+// const base58 = require('bcrypto/lib/encoding/base58');
+const scriptContext = {
+    TOP: 'TOP',
+    P2SH: 'P2SH',
+    P2WPKH: 'P2WPKH',
+    P2WSH: 'P2WSH',
+    P2TR: 'P2TR'
+};
+class Descriptor {
+    constructor(options, network) {
+        this.name = null;
+        this.pubkeys = [];
+        this.subdescriptors = [];
+        this.lastKeyIndex = 0;
+        if (options)
+            this.fromOptions(options, network);
+    }
+    hasPrivateKeys() {
+        for (const key of this.pubkeys) {
+            if (key.getPrivateKey !== null)
+                return true;
+        }
+        return false;
+    }
+    static polyMod(c, val) {
+        const c0 = c.ushr(BN.from(35, 10, 'be'));
+        c = c.and(BN.from('7ffffffff', 16, 'be')).ushl(BN.from(5, 10, 'be')).xor(val);
+        if (c0.and(BN.from(1, 10, 'be')).toNumber())
+            c = c.xor(BN.from('f5dee51989', 16, 'be'));
+        if (c0.and(BN.from(2, 10, 'be')).toNumber())
+            c = c.xor(BN.from('a9fdca3312', 16, 'be'));
+        if (c0.and(BN.from(4, 10, 'be')).toNumber())
+            c = c.xor(BN.from('1bab10e32d', 16, 'be'));
+        if (c0.and(BN.from(8, 10, 'be')).toNumber())
+            c = c.xor(BN.from('3706b1677a', 16, 'be'));
+        if (c0.and(BN.from(16, 10, 'be')).toNumber())
+            c = c.xor(BN.from('644d626ffd', 16, 'be'));
+
+        return c;
+    }
+
+    static createChecksum(desc) {
+        const INPUT_CHARSET =
+            '0123456789()[],\'/*abcdefgh@:$%{}IJKLMNOPQRSTUVWXYZ&+-.;<=>?!^_|~ijklmnopqrstuvwxyzABCDEFGH`#"\\ ';
+        const CHECKSUM_CHARSET = 'qpzry9x8gf2tvdw0s3jn54khce6mua7l';
```

Comment:
These should belong to class and not defined in a function

Reply:
This is used only in createchecksum function. This has nothing to do with descriptor class variables. Descriptor class varables will only be array of pubkeyproviders and array of subdescriptors along with the name of the descriptor.

Reply:
Probably okay. I still don't feel it should be inside this function. Maybe a constant defined separately from the class declaration.
Resolved for now.